File: kmp/composeApp/src/commonMain/kotlin/dev/jianastrero/trainer/data/remote/PokeApiRemote.kt
package dev.jianastrero.trainer.data.remote

import dev.jianastrero.trainer.data.ktor.KtorClient
import dev.jianastrero.trainer.domain.model.pokeapi.response.PokeApiPaginatedResponse
import dev.jianastrero.trainer.domain.model.pokeapi.response.PokemonItemResponse
import dev.jianastrero.trainer.domain.model.pokeapi.response.pokemon.PokemonResponse
import dev.jianastrero.trainer.domain.model.pokeapi.response.species.PokemonSpeciesResponse
import io.ktor.client.call.body
import io.ktor.client.request.get

class PokeApiRemote(
    private val pokeApiClient: KtorClient
) {

    suspend fun getPokemonList(
        offset: Int,
        limit: Int
    ): PokeApiPaginatedResponse<PokemonItemResponse> {
        val response = pokeApiClient.client.get("/api/v2/pokemon?limit=${limit}&offset=${offset}")

        return response.body()
    }

    suspend fun getPokemon(id: String): PokemonResponse {
        val response = pokeApiClient.client.get("/api/v2/pokemon/$id")

        return response.body()
    }

    suspend fun getSpecies(id: String): PokemonSpeciesResponse {
        val response = pokeApiClient.client.get("/api/v2/pokemon-species/$id")

        return response.body()
    }
}

File: react_native/src/data/remote/PokeApiRemote.ts
import axios from "axios";
import {PokemonResponse} from "../../domain/model/pokeapi/response/PokemonResponse.ts";
import PaginatedPokeApiResponse from "../../domain/model/pokeapi/response/PaginatedPokeApiResponse.ts";
import PokemonItemResponse from "../../domain/model/pokeapi/response/PokemonItemResponse.ts";
import PokemonSpeciesResponse from "../../domain/model/pokeapi/response/PokemonSpeciesResponse.ts";

interface PokeApiRemote {
    getAllPokemon: () => Promise<PokemonItemResponse[]>;
    getPokemonById: (pokemonId: string) => Promise<PokemonResponse>;
    getPokemonSpeciesById: (pokemonId: string) => Promise<PokemonSpeciesResponse>;
}

export default function usePokeApiRemote() {
    const getAllPokemon = async (): Promise<PokemonItemResponse[]> => {
        const response = await axios.get<PaginatedPokeApiResponse<PokemonItemResponse>>(`https://pokeapi.co/api/v2/pokemon?offset=0&limit=10000`);
        return response.data.results;
    };
    const getPokemonById = async (pokemonId: string): Promise<PokemonResponse> => {
        const response = await axios.get<PokemonResponse>(`https://pokeapi.co/api/v2/pokemon/${pokemonId}`);
        return response.data;
    }
    const getPokemonSpeciesById = async (pokemonId: string): Promise<PokemonSpeciesResponse> => {
        const response = await axios.get<PokemonSpeciesResponse>(`https://pokeapi.co/api/v2/pokemon-species/${pokemonId}`);
        return response.data;
    }

    return {
        getAllPokemon: getAllPokemon,
        getPokemonById: getPokemonById,
        getPokemonSpeciesById: getPokemonSpeciesById
    } as PokeApiRemote;
};